pkg/agent/httpConfigForCluster.go

```diff
@@ -0,0 +1,112 @@
+	log.Debug("configForCluster.ToDiscoveryClient called")
+	config, err := f.ToRESTConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// The more groups you have, the more discovery requests you need to make.
+	// given 25 groups (our groups + a few custom resources) with one-ish version each, discovery needs to make 50 requests
+	// double it just so we don't end up here again for a while.  This config is only used for discovery.
+	config.Burst = f.discoveryBurst
+
+	cacheDir := filepath.Join(homedir.HomeDir(), ".kube", "cache")
+
+	// retrieve a user-provided value for the "cache-dir"
+	// override httpCacheDir and discoveryCacheDir if user-value is given.
+	if f.CacheDir != nil {
+		cacheDir = *f.CacheDir
+	}
+	httpCacheDir := filepath.Join(cacheDir, "http")
+	discoveryCacheDir := computeDiscoverCacheDir(filepath.Join(cacheDir, "discovery"), config.Host)
+
+	return diskcached.NewCachedDiscoveryClientForConfig(config, discoveryCacheDir, httpCacheDir, time.Duration(10*time.Minute))
+}
+
+// ToRESTMapper requires an implementation on the embedding struct because
+// it calls ToDiscoveryClient.
+func (f *configForCluster) ToRESTMapper() (meta.RESTMapper, error) {
+	log.Debug("configForCluster.ToRESTMapper called")
+	discoveryClient, err := f.ToDiscoveryClient()
+	if err != nil {
+		return nil, err
+	}
+
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(discoveryClient)
+	expander := restmapper.NewShortcutExpander(mapper, discoveryClient)
+	return expander, nil
+}
+
+// ToRawKubeConfigLoader may need to be replicated here, but from limited
+// testing it works just to call the embedded implementation.
+func (f *configForCluster) ToRawKubeConfigLoader() clientcmd.ClientConfig {
+	log.Debug("configForCluster.ToRawKubeConfigLoader called")
+	return f.ConfigFlags.ToRawKubeConfigLoader()
+}
+
+// WithDiscoveryBurst is required because it sets the private var used by
+// ToDiscoveryClient.
+func (f *configForCluster) WithDiscoveryBurst(discoveryBurst int) *configForCluster {
+	f.discoveryBurst = discoveryBurst
+	return f
+}
+
+// The following var and func are only required here because computeDiscoverCacheDir
+// is used in ToDiscoveryClient above.
+
+// overlyCautiousIllegalFileCharacters matches characters that *might* not be supported.  Windows is really restrictive, so this is really restrictive
+var overlyCautiousIllegalFileCharacters = regexp.MustCompile(`[^(\w/\.)]`)
+
+// computeDiscoverCacheDir takes the parentDir and the host and comes up with a "usually non-colliding" name.
+func computeDiscoverCacheDir(parentDir, host string) string {
+	// strip the optional scheme from host if its there:
+	schemelessHost := strings.Replace(strings.Replace(host, "https://", "", 1), "http://", "", 1)
+	// now do a simple collapse of non-AZ09 characters.  Collisions are possible but unlikely.  Even if we do collide the problem is short lived
+	safeHost := overlyCautiousIllegalFileCharacters.ReplaceAllString(schemelessHost, "_")
```

Comment:
why don't you use the `sha` of the host?

Reply:
I didn't want to change this code from the original implementation at all - I just needed to reproduce it here in this implementation, as it's (a private) part of the default implementation for `genericclioptions.ConfigFlags`.